relations/src/shielder/deposit_and_merge.rs

```diff
@@ -41,161 +10,71 @@ use crate::{
-            merkle_path: None,
-            leaf_index: None,
-            old_note: None,
-            old_token_amount: None,
-            new_token_amount: None,
-            _phantom: PhantomData,
-        }
-    }
-}
-
-impl DepositAndMergeRelation<FullInput> {
-    #[allow(clippy::too_many_arguments)]
-    pub fn with_full_input(
-        max_path_len: u8,
-        token_id: FrontendTokenId,
-        token_amount: FrontendTokenAmount,
-        old_nullifier: FrontendNullifier,
-        merkle_root: FrontendMerkleRoot,
-        new_note: FrontendNote,
-        old_trapdoor: FrontendTrapdoor,
-        new_trapdoor: FrontendTrapdoor,
-        new_nullifier: FrontendNullifier,
-        merkle_path: FrontendMerklePath,
-        leaf_index: FrontendLeafIndex,
-        old_note: FrontendNote,
-        old_token_amount: FrontendTokenAmount,
-        new_token_amount: FrontendTokenAmount,
-    ) -> Self {
-        DepositAndMergeRelation {
-            max_path_len,
-            // Public inputs
-            token_id: Some(BackendTokenId::from(token_id)),
-            token_amount: Some(BackendTokenAmount::from(token_amount)),
-            old_nullifier: Some(BackendNullifier::from(old_nullifier)),
-            merkle_root: Some(BackendMerkleRoot::from(BigInteger256::new(merkle_root))),
-            new_note: Some(BackendNote::from(BigInteger256::new(new_note))),
-            // Private inputs.
-            old_trapdoor: Some(BackendTrapdoor::from(old_trapdoor)),
-            new_trapdoor: Some(BackendTrapdoor::from(new_trapdoor)),
-            new_nullifier: Some(BackendNullifier::from(new_nullifier)),
-            merkle_path: Some(
-                merkle_path
-                    .iter()
-                    .map(|node| BackendNote::from(BigInteger256::new(*node)))
-                    .collect(),
-            ),
-            leaf_index: Some(BackendLeafIndex::from(leaf_index)),
-            old_note: Some(BackendNote::from(BigInteger256::new(old_note))),
-            old_token_amount: Some(BackendTokenAmount::from(old_token_amount)),
-            new_token_amount: Some(BackendTokenAmount::from(new_token_amount)),
-
-            _phantom: PhantomData,
-        }
+#[snark_relation]
+mod relation {
+    use core::ops::Add;
+
+    use ark_r1cs_std::{alloc::AllocVar, eq::EqGadget, fields::fp::FpVar, ToBytesGadget};
+    use ark_relations::ns;
+
+    use crate::shielder::{
+        check_merkle_proof, convert_hash, convert_vec,
+        note::check_note,
+        types::{
+            BackendLeafIndex, BackendMerklePath, BackendMerkleRoot, BackendNote, BackendNullifier,
+            BackendTokenAmount, BackendTokenId, BackendTrapdoor, FrontendLeafIndex,
+            FrontendMerklePath, FrontendMerkleRoot, FrontendNote, FrontendNullifier,
+            FrontendTokenAmount, FrontendTokenId, FrontendTrapdoor,
+        },
+    };
+
+    #[relation_object_definition]
+    struct DepositAndMergeRelation {
+        #[constant]
+        pub max_path_len: u8,
```

Comment:
could we avoid shorthanded variable names? They serve no purpose, can be ambiguous, and can make code harder to follow...

Reply:
No sure if I follow. You propose to remove the variable or change it's name to `maximal_path_length`?

Reply:
i propose using full, non-abbreviated, human readable names - `maximal_path_length`

Reply:
I agree with the below issue with `circuit_without_input` however here we use standards shorthands that could well replace the long versions since they are so widely used and understood. I'd replace the below one according to your suggestion but keep this one.